Add an option to highlight Hard Beats

Snootiful opened the pull request:
## Why I opened this PR

I've been playing with tau for some time now, and it bothered me somewhat that the hard beats have the same color as regular beats. This makes it a bit harder for me to differentiate between the two and kinda makes them blur together.

When playing with the Strict mod this gets even worse. There — because the hard beats start off quite thin — it's very difficult to differentiate between hard beats and regular beats as they look almost the same.

<img src="https://github.com/taulazer/tau/assets/73030116/2022c827-b4ea-4067-97a6-92c17e5e97f1" width="40%" title="Approaching hard beat near the center of the play area">

To alleviate this I made this PR, which adds an option that changes the color of hard beats. My goal is to improve usability for players with color blindness or impaired vision, and players with some neurodevelopmental disorders.[^1]


## Changes

With the option "Highlight Hard Beats" enabled, all hard beats will now be colored orange. This includes regular hard beats, strict hard beats, and strict hard beat sliders.
I picked orange for the time being since it seems to work well enough with the accent color. It also looks different enough from white to make it easier to distinguish, but it could be improved upon.

<img src="https://github.com/taulazer/tau/assets/73030116/9c8cd7b6-7d4f-4e52-83d7-8c54d603e2a8" width="40%" title="Highlighted hard beat with a few regular beats and a slider">
<img src="https://github.com/taulazer/tau/assets/73030116/7462203a-5700-4bad-a713-c6d1c10da526" width="40%" title="Highlighted hard beats and sliders with the Strict mod">
<img src="https://github.com/taulazer/tau/assets/73030116/abdfde1d-558f-4afc-9b83-af3e0d57fff2" title="Highlight hard beats checkbox in the settings menu">

The relevant tests have been updated with additional steps to verify that these changes work as expected. Manual tests with a release build of osu! have been conducted as well.

The cached slider texture from `TauCachedProperties` has been removed to allow for changing the color of the hard beat sliders. It could be reintroduced in some fashion if it turns out these changes have a noticeable impact on performance. By far the easiest way to do this would be to revert the changes to `TauCachedProperties` and `DrawableSlider`.


## Notes

Because I'm not very familiar with C# and don't use it often the code I deliver might not be very tidy, and there will undoubtedly be inefficiencies and strange decisions all over. I did however attempt to match the style of the code I was working with.
If there is anything I could improve upon, please let me know.

[^1]: I have Autism without any visual impairments, yet I can tell that my ability to read certain patterns in both tau and standard really struggles compared to other players. Of course, I can't speak for other people with neurodevelopmental disorders, color blindness, or visual impairments.

LumpBloom7 commented on osu.Game.Rulesets.Tau/Localisation/SettingStrings.cs:
```diff
@@ -26,6 +26,11 @@ public static class SettingStrings
+        /// <summary>
+        /// "Highlight Hard Beats"
+        /// </summary>
+        public static LocalisableString HighlightHardBeats = new TranslatableString(getKey(@"highlight_hard_beats"), @"Highlight Hard Beats");
```
Add the corresponding string into Settings.resx too

Snootiful replied:
Would this string need to be added to all Settings.*.resx files or just Settings.resx?

LumpBloom7 replied:
Just Settings.resx, crowdin manages the rest

LumpBloom7 commented on osu.Game.Rulesets.Tau/Objects/Drawables/DrawableSlider.cs:
```diff
@@ -159,6 +172,7 @@ protected override void OnApply()
         {
             base.OnApply();
             path.FadeColour = colour.ForHitResult(HitResult.Great);
+            HighlightHardBeats.BindValueChanged(value => path.Texture = value.NewValue && HitObject.IsHard ? pathTextureHighlighted : pathTexture, true);
```
The setting is not dependent on the hitobject, so this does not need to be done on every time this drawable is assigned to a HitObject. This can be done once in `load()`.

Snootiful replied:
Disregard, solved the problem by overriding `LoadComplete()`. Binding in `load()` then calling `TriggerChange()` on the bound property allows me to use `renderer` for generating the textures inside the handler while allowing `HitObject` to be populated.

LumpBloom7 commented on osu.Game.Rulesets.Tau/Objects/Drawables/DrawableSlider.cs:
```diff
@@ -176,6 +190,7 @@ protected override void OnFree()
             base.OnFree();
             trackingCheckpoints.Clear();
             slidingSample?.ClearSamples();
+            HighlightHardBeats.UnbindEvents();
```
Same as the above.

LumpBloom7 commented on osu.Game.Rulesets.Tau/Objects/Drawables/Pieces/HardBeatPiece.cs:
```diff
@@ -1,21 +1,24 @@
             Masking = true;
             BorderThickness = 5;
-            BorderColour = Color4.White;
+            BorderColour = HighlightHardBeats.Value ? Color4.Orange : Color4.White;
```
This will always be false, and is redundant.

The correct colour will be assigned when the config is bound in load (via `BindValueChanged`)

LumpBloom7 commented on osu.Game.Rulesets.Tau/Objects/Drawables/Pieces/StrictHardBeatPiece.cs:
```diff
@@ -12,10 +13,11 @@ public partial class StrictHardBeatPiece : CircularProgress
         public StrictHardBeatPiece()
         {
-            Colour = Color4.White;
+            Colour = HighlightHardBeats.Value ? Color4.Orange : Color4.White;
```
Same as the comment on `HardBeatPiece`

LumpBloom7 commented on osu.Game.Rulesets.Tau/Objects/Drawables/Pieces/HardBeatPiece.cs:
```diff
@@ -30,6 +33,13 @@ public HardBeatPiece()
+        }
+
+        [BackgroundDependencyLoader]
+        private void load(TauRulesetConfigManager config)
```
Prolly better to mark this as nullable (`TauRulesetConfigManager?`) since you treat is as nullable underneath.

LumpBloom7 commented on osu.Game.Rulesets.Tau/UI/TauCachedProperties.cs:
```diff
@@ -23,10 +21,6 @@ public void SetRange(float cs)
-            SliderTexture?.Dispose();
-            SliderTexture = null;
-        }
+        public void Dispose() {}
```
Since there is no longer anything to `Dispose`, this method can be removed, and this class doesn't need to be an `IDisposable`

LumpBloom7 commented on osu.Game.Rulesets.Tau/Objects/Drawables/DrawableSlider.cs:
```diff
@@ -159,6 +172,7 @@ protected override void OnApply()
         {
             base.OnApply();
             path.FadeColour = colour.ForHitResult(HitResult.Great);
+            HighlightHardBeats.BindValueChanged(value => path.Texture = value.NewValue && HitObject.IsHard ? pathTextureHighlighted : pathTexture, true);
```
Also put parenthesis around the ternary for readability.

LumpBloom7 commented on osu.Game.Rulesets.Tau/Objects/Drawables/DrawableSlider.cs:
```diff
@@ -141,13 +147,20 @@ protected override void ClearNestedHitObjects()
 
             host.DrawThread.Scheduler.AddDelayed(() => drawCache.Invalidate(), 0, true);
-            path.Texture = properties.SliderTexture ??= generateSmoothPathTexture(renderer, path.PathRadius, _ => Color4.White);
+            pathTexture = generateSmoothPathTexture(renderer, path.PathRadius, _ => Color4.White);
+            pathTextureHighlighted = generateSmoothPathTexture(renderer, path.PathRadius, _ => Color4.Orange);
```
It seems odd to always generate both textures at once, when the most people aren't toggling the setting during gameplay.

Would it be feasible to only generate the texture only when it is needed? Something along the lines of the following, but cleaner ofc.
```cs
HighlightHardBeats.BindValueChanged(value => path.Texture = ((value.NewValue && HitObject.IsHard) ? 
    pathTextureHighlighted ??= generateSmoothPathTexture(renderer, path.PathRadius, _ => Color4.Orange) : 
    pathTexture ??= generateSmoothPathTexture(renderer, path.PathRadius, _ => Color4.White);, true));
```

LumpBloom7 reviewed (changes requested):
The option to "Highlight Hard Beats" should not also highlight sliders. Prolly better to call the option something like "Increase visual distinction" or something that is doesn't suggest that it applies to one of the two objects.

Not sure about the performance impact of removing the use of the cache, I'd reckon there is a reason it was used. Maybe someone else can corroborate? (Normally we don't do stuff like that unless the impact is significant)

Snootiful commented:
Updated the code following the above instructions.
I was able to use the cached properties again by adding an additional Texture for the visually distinct slider body.
Also I renamed the option and it's variables/properties anyway to be clearer, since I like "Improve Visual Distinction" better.

naoei commented:
I don't like the direction of turning this into a setting, Tau already has too many settings as-is and some settings should be ideally removed (i.e. Kiai effects).

This current implementation locks the user behind one colour (orange), and if say someone were colourblind or has some other visual impairments that would blend red and orange together, this setting would not help at all. The ideal implementation would be to allow the user to change any hit object that would require the hard beat press to any colour they want via skinning. We did have an [earlier implementation of skinning](https://github.com/taulazer/tau/pull/140), but it was very rudimentary and I'm currently thinking on how we can do better.

There's also the possibility that we might not want this feature in the future, I have a development branch on my fork that tests to see if we would want to do simultaneous button presses (only Z/X) for hard beats instead of a separate button for them.

naoei commented:
Suspending this for now for previous concerns.
